### PR TITLE
Siri Hemsley & Anna Lenksjö - Project Auth

### PR DESCRIPTION
**Pull request by sirimatsdotter**
**Link to my deployed project:**
Frontend: https://youthful-liskov-e3177b.netlify.app/
Backend: https://siri-anna-auth-backend.herokuapp.com

**My reflections on how this week's project turned out:**

- Overall it went well and it was fun to do our first real fullstack application!
- It was great to refresh our knowledge on so many different things, such as React Router, Redux Store.
- Really good to learn about authorization and how to create logins etc

**Things I'd like to have clarified or explained in more detail:**

- As mentioned in the README we do still have one error which we don't know exactly what it means. Would be great to get some more explanation on this :)

_"index.js:1 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
in LogIn (at App.js:43)"_

### CODE REVIEW

File: backend/server.js [hunk 1]
```diff
@@ -1,26 +1,122 @@
...
+      'content-type': 'application/json; charset=utf-8',
+    }
+  }
+  await axios(config)
+    .then((response) => {
+      const images = response.data
+      res.send({success: true, images})
+    })
+    .catch((err) => res.status(401).json({ success: false, message: 'Not authorized', err }))
+})
```
Comment (idasylven) [inline backend/server.js:73]: Cool that you used axios! 😎

File: frontend/src/App.js [hunk 2]
```diff
@@ -1,9 +1,56 @@
 import React from 'react'
+import { Provider } from 'react-redux'
+import {configureStore, combineReducers } from '@reduxjs/toolkit'
+import { BrowserRouter, Switch, Route } from 'react-router-dom'
+import styled from 'styled-components'
```
Comment (idasylven) [inline frontend/src/App.js:5]: Great job using styled components! 👍🏻  Perhaps you could consider using macro too, if you wish for the class names to be displayed in dev tools ⚒️

File: frontend/src/components/StyledBtn.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  margin: 10px;
+  background-color: #CAEBF2;
+  border: 1px solid #9099A5;
+  border-radius: 2px;
+  cursor: pointer;
+  box-shadow: 7px 7px white, 7px 7px 0px 1px #9099A5;
+  &:hover {
+    background-color: rgb(203,200,254);
+  }
+`
```
Comment (idasylven) [inline frontend/src/components/StyledBtn.js:17]: In the Sign Up or Log In form, perhaps you could consider adding some margin to the bottom of the button or padding to the bottom of the form. When an error message is displayed, it's getting a little cramped ➡️ ⬅️

**Review by idasylven — APPROVED** [review]
Great job on this assignment! Clean and well-structured code, and we get the impression you really fully understand what you're doing 💥 🍾 

Nice README explaining your process and what you would like to improve in the future, with more specific error messages and the logged in site to be a bit more complex 👍🏻 😻

We have very few improving comments except for being able to sign up without a password, and also being able to access the secret content without a password. Perhaps you want to look into that 👀. (And let us in on the solution 😝 )

APPROVED! ✨

File: backend/server.js [hunk 4]
```diff
@@ -1,26 +1,122 @@
...
+      username,
+      password: bcrypt.hashSync(password, salt)
+     }).save()
+
+    res.json({
+      success: true,
+      userID: newUser._id,
+      username: newUser.username,
+      accessToken: newUser.accessToken 
+     })
```
Comment (KarolCislowski) [inline backend/server.js:90]: 👍🏻

File: frontend/src/App.js [hunk 5]
```diff
@@ -1,9 +1,56 @@
...
+display: flex;
+flex-direction: column;
+height: 100vh;
+width: 100vw;
+justify-content: center;
+align-items: center;
+background-image: linear-gradient(#CBC7FE, #B9DEFF);
+box-sizing: border-box;
+font-family: "Now Bold";
+color: #6A7885;
```
Comment (KarolCislowski) [inline frontend/src/App.js:24]: indentation

File: frontend/src/components/Content.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      dispatch(user.actions.setErrors(null))
+    })
+    localStorage.clear()
+    history.push('/')
+  }
+
+  return (
+    <>
+    {imgsrc ? <Image src={imgsrc} alt="cat" /> : <p>Loading image...</p>}
+    <StyledButton onClick={onLogOut}>Log out</StyledButton>
```
Comment (KarolCislowski) [inline frontend/src/components/Content.js:63]: indentation

File: frontend/src/components/Content.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      .then(res => res.json())
+      .then(data => {
+        if (data.success) {
+          setImgsrc(data.images[0].url)
+        } else {
+          dispatch(user.actions.setErrors(data))
+        }
+      })
+    }
+  },[accessToken, dispatch])
```
Comment (KarolCislowski) [inline frontend/src/components/Content.js:48]: `useEffect` hook should be always at the root of your component. Never nest it in a different method even the other `useEffect`. 

```jsx

useEffect(() => {
  if (!accessToken) {
    history.push('/')
  } else {
    const options = {
      method: 'GET',
      headers: {
        Authorization: accessToken
      }
    }
    fetch(API_URL('session'), options)
      .then(res => res.json())
      .then(data => {
        if (data.success) {
          setImgsrc(data.images[0].url)
        } else {
          dispatch(user.actions.setErrors(data))
        }
      })
  }
}, [accessToken, dispatch])


```

File: frontend/src/reducers/user.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    }, 
+    setAccessToken: (store, action) => {
+      store.accessToken = action.payload
+      window.localStorage.setItem('accessToken', action.payload)
+    },
+    setErrors: (store, action) => {
+      store.errors = action.payload
+    }
+  }
+})
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:24]: great reducer 👍🏻

File: frontend/src/components/SignUp.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      },
+      body: JSON.stringify({
+        username,
+        password
+      })
+    })
+      .then(response => response.json())
+      .then(data => {
+        if (data.success) {
+          history.push('/login')
```
Comment (KarolCislowski) [inline frontend/src/components/SignUp.js:59]: Why? Your API already provided you with all credentials here so why not log your user behind the scenes and redirect straight to the content?

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
It looks really nice but you need to do some fixes:
1. for sure you need to kill the Netlify error interruption issue, instruction is in the comment with screenshot. 
2. Analise differences between my and your useEffect approach and apply it to your code now and for future projects 😜 


Because your graduation is so soon please let me know on slack (a short note with a link to this PR) if either you finish it or still have the error you mentioned, then I will investigate it with you after Netlify will not interrupt us 🕵🏻

File: pull_request_template.md [hunk 10]
```diff
@@ -1,5 +1,16 @@
...
+- Overall it went well and it was fun to do our first real fullstack application!
+- It was great to refresh our knowledge on so many different things, such as React Router, Redux Store.
+- Really good to learn about authorization and how to create logins etc
+
 Things I'd like to have clarified or explained in more detail:
+
+- As mentioned in the README we do still have one error which we don't know exactly what it means. Would be great to get some more explanation on this :)
+
+"index.js:1 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
+in LogIn (at App.js:43)"
```
Comment (KarolCislowski) [inline pull_request_template.md:16]: To fix it you need to create `_redirects` file in the `public` directory of your react project and paste it inside:
``` 
/* /index.html 200
```

image

**Review by KarolCislowski — CHANGES_REQUESTED** [review]

**Review by KarolCislowski — APPROVED** [review]
Great Project 🥇

**Pull request closed without merging**